### PR TITLE
[ESI] Global services model

### PR DESCRIPTION
**Pull request by teqdruid**
ESI global services provide connectivity to chip-level stuff like PCIe, DRAM, Ethernet, etc. Additionally, services are completely user-defined and can be used recursively, so a user could define a telemetry service which is backed by PCIe (for instance). Users need to implement the services themselves, but ESI will plumb through the connection requests and provide users with a list of pre-connected ESI ports. ESI could also provide several standard multiplexing components.

Implements roughly the model in #2811. A lowering pass as described in the op documentation will be in a subsequent PR.

### CODE REVIEW

**Comment by teqdruid** [comment]
FYI @mikeurbach @mortbopet @makslevental

**Comment by mikeurbach** [comment]
👍  got it, just checking. I was seeing some code sharing related to some of the older MSFTPartition work, and I wasn't totally sure from a quick skim if this was "port punching" or pulling everything into a new "partition".

**Comment by teqdruid** [comment]
I just got the "port punching" working last night, so I haven't written any tests or code docs yet. The code sharing is really only for the code to sort the module instantiation graph by use. I refactored the partition code in such a way that it's not appropriate to share it, though the new `bubbleup` ESI code looks quite similar.

**Comment by teqdruid** [comment]
Including the service model and connection tunneling pass in one PR was too much. Separating out the model.

**Comment by teqdruid** [comment]
The `esi.service.instance` regular I/O ports should really be more like `hw.instance` ports (i.e. with names), but until we refactor the `hw.instance` parsing/printing code it's just too much effort.

File: test/Dialect/ESI/services.mlir [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// RUN: circt-opt %s | circt-opt | FileCheck %s
```
Comment (mortbopet) [inline test/Dialect/ESI/services.mlir:1]: Suggesting renaming this file to `services_round_trip.mlir` to be consistent with naming used in other dialect tests.

File: include/circt/Dialect/ESI/ESIServices.td [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    implements from the containing modules' descendants (in the instance
+    hierarchy). It bubbles them all up to the module containing this op,
+    creating the necessary ESI channel ports, groups them appropriately, then
+    replaces this op with a `service.impl_req`.
+
+    Since implementing the server will usually need "normal" I/O, `inputs` and
+    `results` act like normal `hw.instance` ports.
+
+    $identifier is used by frontends to specify or remember the type of
+    implementation to use for this service.
```
Comment (mortbopet) [inline include/circt/Dialect/ESI/ESIServices.td:73]: I think i understand the intention here, but i find the IR a bit difficult to reason about (`%c = esi.service.instance "topComms" @HostComms (%clk) : (i1) -> (i8)`) in terms of what's doing what between `"topComms"` and `@HostComms`. Could the op format be modified to better reflect the intention of the attribute?
e.g.:
```mlir
%c = esi.service.instance @HostComms impl as "topComms" (%clk) : (i1) -> (i8)
```

File: lib/Dialect/ESI/ESIOps.cpp [hunk 3]
```diff
@@ -204,5 +204,57 @@ LogicalResult UnwrapSVInterface::verify() {
...
+  if (!serviceDeclOp) {
+    op->emitOpError("Cannot find module ") << modName;
+    return {};
+  }
+
+  StringAttr innerSym = servicePort.getName();
+  for (auto portDecl : serviceDeclOp.getOps<OpType>())
+    if (portDecl.inner_symAttr() == innerSym)
+      return portDecl;
+  op->emitOpError("Cannot find port named ") << innerSym;
```
Comment (mortbopet) [inline lib/Dialect/ESI/ESIOps.cpp:226]: Missing test

File: test/Dialect/ESI/errors.mlir [hunk 4]
```diff
@@ -43,3 +43,35 @@ hw.module @test(%m : !sv.modport<@IData::@Noexist>) {
...
+// -----
+
+esi.service.decl @HostComms {
+  esi.service.to_server @Send : !esi.channel<i16>
+  esi.service.to_client @Recv : !esi.channel<i32>
+}
+
+hw.module @Loopback (%clk: i1) -> () {
+  %dataIn = esi.service.req.to_client <@HostComms::@Recv> (["loopback_tohw"]) : !esi.channel<i32>
+  // expected-error @+1 {{Request type does not match port type '!esi.channel<i16>'}}
```
Comment (mortbopet) [inline test/Dialect/ESI/errors.mlir:56]: Checks for `EmitOpError`s should also match on the op name in the error output, to ensure that the error is attached to the right op. e.g.:
```
// expected-error @+1 {{'esi.service.req.to_server' op Request type does not match port type '!esi.channel<i16>'}}
```

File: test/Dialect/ESI/errors.mlir [hunk 5]
```diff
@@ -43,3 +43,35 @@ hw.module @test(%m : !sv.modport<@IData::@Noexist>) {
...
+
+
+// -----
+
+esi.service.decl @HostComms {
+  esi.service.to_client @Recv : !esi.channel<i8>
+}
+
+hw.module @Loopback (%clk: i1) -> () {
+  // expected-error @+1 {{Request type does not match port type '!esi.channel<i8>'}}
```
Comment (mortbopet) [inline test/Dialect/ESI/errors.mlir:68]: as above.

File: test/Dialect/ESI/errors.mlir [hunk 6]
```diff
@@ -43,3 +43,35 @@ hw.module @test(%m : !sv.modport<@IData::@Noexist>) {
...
+  %dataIn = esi.service.req.to_client <@HostComms::@Recv> (["loopback_tohw"]) : !esi.channel<i32>
+}
+
+// -----
+
+hw.module @Loopback (%clk: i1) -> () {
+  // expected-error @+1 {{Cannot find module "HostComms"}}
```
Comment (mortbopet) [inline test/Dialect/ESI/errors.mlir:75]: as above.

**Pull request merged**